sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -413,9 +413,13 @@ case class DataSource(
         } else {
           val globbedPaths = checkAndGlobPathIfNecessary(
             checkEmptyGlobPath = true, checkFilesExist = checkFilesExist)
-          val index = createInMemoryFileIndex(globbedPaths)
+          val fileStatusCache = FileStatusCache.getOrCreate(sparkSession)
+          val indexInSchemaInferring = new InMemoryFileIndex(
+            sparkSession, globbedPaths, options, userSpecifiedSchema, fileStatusCache)
           val (resultDataSchema, resultPartitionSchema) =
-            getOrInferFileFormatSchema(format, () => index)
+            getOrInferFileFormatSchema(format, () => indexInSchemaInferring)
+          val index = new InMemoryFileIndex(
+            sparkSession, globbedPaths, options, Some(resultPartitionSchema), fileStatusCache)
```

Comment:
New index re-uses the file statuses cache `fileStatusCache`, so, this should allow to avoid additional accesses to the file  system.